src/metabase/analytics/prometheus.clj

```diff
@@ -0,0 +1,248 @@
+                  (setting/get-raw-value :prometheus-server-port integer? parse))))
+
+(defonce ^:private ^{:doc "Prometheus System for prometheus metrics"} system nil)
+
+(p.types/defprotocol+ PrometheusActions
+  (stop-web-server [this]))
+
+(p/defrecord+ PrometheusSystem [registry web-server]
+  ;; prometheus just runs in the background collecting metrics and serving them from
+  ;; localhost:<prometheus-server-port>/metrics. Nothing we need to do but shutdown.
+  PrometheusActions
+  (stop-web-server [_this]
+    (when-let [^Server web-server web-server]
+      (.stop web-server))))
+
+(declare setup-metrics! start-web-server!)
+
+(defn- make-prometheus-system
+  "Takes a port (zero for a random port in test) and a registry name and returns a [[PrometheusSystem]] with a registry
+  serving metrics from that port."
+  [port registry-name]
+  (try
+    (let [registry (setup-metrics! registry-name)
+          web-server (start-web-server! port registry)]
+      (->PrometheusSystem registry web-server))
+    (catch Exception e
+      (throw (ex-info (trs "Failed to initialize Prometheus on port {0}" port)
+                      {:port port}
+                      e)))))
+
+;;; Collectors
+
+(defn c3p0-stats
+  "Takes `raw-stats` from [[metabase.troubleshooting/connection-pool-info]] and groups by each property type rather than each database.
+  {\"metabase-postgres-app-db\" {:numConnections 15,
+                                 :numIdleConnections 15,
+                                 :numBusyConnections 0,
+                                 :minPoolSize 1,
+                                 :maxPoolSize 15},
+   \"db-2-postgres-clean\" {:numConnections 2,
+                            :numIdleConnections 2,
+                            :numBusyConnections 0,
+                            :minPoolSize 1,
+                            :maxPoolSize 15}}
+  Becomes {:numConnections [{:name :numConnections,
+                             :value 15.0, ;; values are all doubles
+                             :timestamp 1662563931039,
+                             :label \"metabase-postgres-app-db\"}
+                            {:name :numConnections,
+                             :value 2.0,
+                             :timestamp 1662563931039,
+                             :label \"db-2-postgres-clean\"}]
+          ...}"
+  [raw-stats]
+  (let [now    (.toEpochMilli (java.time.Instant/now))
+        sample (fn sample [[db-label k v]]
+                 {:name      k
+                  :value     (double v)
+                  :timestamp now
+                  :label     db-label})]
+    (->> raw-stats
+         (mapcat (fn [[db-label values]]
+                   (map (fn [[k v]] [db-label k v]) values)))
+         (map sample)
+         (group-by :name))))
+
+(def ^:private label-translation
+  {:maxPoolSize        {:label       "c3p0_max_pool_size"
+                        :description "C3P0 Max pool size"}
```

Comment:
Would it make sense to support translations here?

Reply:
oh yeah. maybe so.

Reply:
I'm doing it on the description. I think we want the keys to always remain the same?

Reply:
Yes, I think the keys should stay the same.